src/plugins/messageLatency/index.tsx

```diff
@@ -0,0 +1,147 @@
+import { Devs } from "@utils/constants";
+import { isNonNullish } from "@utils/guards";
+import definePlugin, { OptionType } from "@utils/types";
+import { waitFor } from "@webpack";
+import { SnowflakeUtils, Tooltip } from "@webpack/common";
+import { Message } from "discord-types/general";
+
+type FillValue = ("status-danger" | "status-warning" | "text-muted");
+type Fill = [FillValue, FillValue, FillValue];
+type DiffKey = keyof Diff;
+
+interface Diff {
+    days: number,
+    hours: number,
+    minutes: number,
+    seconds: number;
+}
+
+let HiddenVisually;
+waitFor("HiddenVisually", mod => {
+    HiddenVisually = mod.HiddenVisually;
+});
+
+export default definePlugin({
+    name: "MessageLatency",
+    description: "Displays an indicator for messages that took ≥n seconds to send",
+    authors: [Devs.arHSM],
+    settings: definePluginSettings({
+        latency: {
+            type: OptionType.NUMBER,
+            description: "Threshold in seconds for latency indicator",
+            default: 2
+        }
+    }),
+    patches: [
+        {
+            find: "showCommunicationDisabledStyles",
+            replacement: {
+                match: /(message:(\i),avatar:\i,username:\(0,\i.jsxs\)\(\i.Fragment,\{children:\[)(\i&&)/,
+                replace: "$1$self.Tooltip($2),$3"
+            }
+        }
+    ],
+    stringDelta(delta: number) {
+        const diff: Diff = {
+            days: Math.round(delta / (60 * 60 * 24)),
+            hours: Math.round((delta / (60 * 60)) % 24),
+            minutes: Math.round((delta / (60)) % 60),
+            seconds: Math.round(delta % 60),
+        };
+
+        const str = (k: DiffKey) => diff[k] > 0 ? `${diff[k]} ${k}` : null;
+        const keys: DiffKey[] = ["days", "hours", "minutes", "seconds"];
+
+        return keys.map(str).filter(isNonNullish).join(" ");
+    },
+    latencyTooltipData(message: Message) {
+        const { id, nonce } = message;
+
+        // Message wasn't received through gateway
+        if (!isNonNullish(nonce)) return null;
+
+        const delta = Math.round((SnowflakeUtils.extractTimestamp(id) - SnowflakeUtils.extractTimestamp(nonce)) / 1000);
+
+        // Thanks dziurwa (I hate you)
+        // This is when the user's clock is ahead
+        // Can't do anything if the clock is behind
+        const abs = Math.abs(delta);
+        const ahead = abs !== delta;
+
+        const stringDelta = this.stringDelta(abs);
+
+        // Also thanks dziurwa
+        // 2 minutes
+        const TROLL_LIMIT = 2 * 60;
+        const { latency } = this.settings.store;
+
+        const fill: Fill = delta >= TROLL_LIMIT || ahead ? ["text-muted", "text-muted", "text-muted"] : delta >= (latency * 2) ? ["status-danger", "text-muted", "text-muted"] : ["status-warning", "status-warning", "text-muted"];
+
+        return abs >= latency ? { delta: stringDelta, ahead: abs !== delta, fill } : null;
+    },
+    Tooltip(message: Message) {
```

Comment:
please wrap this in a noop-ErrorBoundary to avoid crashing in the worst case scenario

```suggestion
    Tooltip: ErrorBoundary.wrap(({ message }: { message: Message }) => {
```

(you'll also have to slightly adjust ur patch since now it expects a proper props object as first arg)